_source/_posts/2023-07-28-oidc_workshop.md

```diff
@@ -0,0 +1,505 @@
+
+The frontend is now ready to use the backend's `/api/openid/check` endpoint to investigate whether a user belongs to an OIDC org, and then hit the org's custom `/api/openid/start/${org_id}` endpoint to initiate the login flow if an org is found! Your next step is to add these endpoints to the application's backend. 
+
+# Backend Changes
+
+The application's backend will use a library to handle most of the OIDC authentication process. After authenticating, the user will be redirected back to the app just as they are when logging in with a password.
+
+## Install Passport OIDC Library
+
+Since the sample app is using Passport for authentication, the Passport OIDC library will offer seamless integration. 
+
+In your project, run `npm install --save passport-openidconnect` to install the library.  This will update `package.json` automatically to specify the installed version: 
+
+```
+    "passport-openidconnect": "^0.1.1",
+```
+
+## Add Helper Functions
+
+Add the following helper functions to `apps/api/src/main.ts` to simplify your upcoming work. 
+
+`orgFromId` will take an integer org ID and return the first database result for an organization that the ID references:
+
+```
+async function orgFromId(id) {
+  const org = await prisma.org.findFirst({
+    where: {
+      id: parseInt(id)
+    }
+  })
+  return org
+}
+```
+`getDomainFromEmail` will parse a domain from an email address, by returning everything after the `@`: 
+
+```
+function getDomainFromEmail(email) {
+  let domain;
+  try {
+    domain = email.split('@')[1];
+  } catch(e) {
+    return null;
+  }
+  return domain;
+}
+```
+
+## Add OpenID Org Check Route
+
+The `api/openid/check` endpoint will return the numeric ID of the org that a user's email domain belongs to, or `null` if the user isn't part of an org. The user's full email address will arrive at the endpoint in the request body, so the helper function `getDomainFromEmail` will return the domain, before it's used to look up the corresponsding org in the database. Add this code to `main.ts`: 
+
+```
+app.post('/api/openid/check', async (req, res, next) => {
+  const { username } = req.body;
+
+  const domain = getDomainFromEmail(username);
+  if(domain) {
+    var org = await prisma.org.findFirst({
+      where: {
+        domain: domain
+      }
+    });
+    if(!org) {
+      org = await prisma.org.findFirst({
+        where: {
+          User: {
+            some: {
+              email: username
+            }
+          }
+        }
+      })
+    }
+    if(org && org.issuer) {
+      return res.json({ org_id: org.id });
+    }
+  }
+
+  res.json({ org_id: null });
+});
+```
+
+If the user belongs to an org and the database knows the `issuer`, or OIDC server, for that org, the org's ID will be returned. Otherwise, a `null` result will indicate that the user is not part of an org that the app knows about. 
```

Comment:
```suggestion
If the user belongs to an org and the database knows the `issuer` (the OIDC server) for that org, the org's ID will be returned. Otherwise, a `null` result will indicate that the user is not part of an org that the app knows about. 
```